File: assignments/079-Minesweeper_cpp/minesweeper.cpp

#include <assert.h>
#include <limits>

#include <iostream>
#include <iomanip>
using namespace std;

enum ClickType {CLICK_KNOWN_MINE, CLICK_INVALID, CLICK_CONTINUE, CLICK_LOSE };

#define KNOWN_MINE -3
#define HAS_MINE -2
#define UNKNOWN -1

int readInteger() {
    int value;
    while (!(std::cin >> value)) {
        if (std::cin.eof()) {
            cerr << "End of file from keyboard reading a number.  Quitting\n";
            exit(EXIT_FAILURE);
        }
        
        std::cin.clear(); // Clear the error flag
        std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n'); // Discard the invalid input
        std::cout << "You did not enter any valid number\n" <<
                 "Please try again\n";
    }
    
    return value;
}

class Square {
private:
    int square = UNKNOWN;
    
public:
    Square() {}

    void assign_mine()       { square = HAS_MINE;    }
    void setCount(int count) { square = count;  }
    void mark_known_mine()   { square = KNOWN_MINE;}

    bool is_mine()        { return (square == HAS_MINE || square == KNOWN_MINE);  }
    bool known_mine()     { return square == KNOWN_MINE;   }
    bool has_mine()       { return square == HAS_MINE; }
    bool is_unknown()  { return square == UNKNOWN; }
    int count() {return square; }

   ClickType click (int surroundingMineCount) {

      if (known_mine()) {  // don't allow the player to kill themselves
          return CLICK_KNOWN_MINE;
      }
      if (has_mine()) {
          return CLICK_LOSE;
      }
      if (square >= 0) {  // they've already know this answer
          return CLICK_CONTINUE;
      }
      // no mine here, so show the player the number of mines around this square
      setCount(surroundingMineCount);

      return CLICK_CONTINUE;
  }

};

ostream &operator<<(ostream &out, Square s) {
    if      (s.known_mine())                 { cout << "*"; }  
    else if (s.has_mine() or s.is_unknown()) { cout << "?"; }
    else {
        int mine_count = s.count();
        if (mine_count == 0) { cout << " ";   }
        else { cout << mine_count; }
    }
    return out;
}


class Board {
private:
    Square **board;
    const int width;
    const int height;
    const int totalMines;

    bool in_bounds(int x, int y) {
        return x >= 0 && x < width && y >= 0 && y < height;
    }

public:
  Board(int w, int h, int numMines): width{w}, height{h}, totalMines{numMines} {
      /* write code*/
  }

  ~Board() {
      /* write code*/
  }

  void addRandomMine() {
      int x;
      int y;
      // we could have a board too small for the number of requested mines.
      // try w*h*10 times before giving up
      int limit = width * height * 10;
      do {
          x = random() % width;
          y = random() % height;
          limit--;
      } while(board[y][x].has_mine() && limit > 0);
      board[y][x].assign_mine();
  }


  int countMines(int x, int y) {
      /* write code*/
  }

  bool checkWin() {
      /* write code*/
  }

  int countKnownMines() {
      int count = 0;
      for (int x =0; x < height; x++) {
          for (int y = 0; y < width; y++) {
              if (board[x][y].known_mine()) {
                  count++;
              }
          }
      }
      return count;
  }
	
  void printBoard() {
      cout << "    ";
      for (int x = 0; x < width; x++) { cout << x/10;  } cout << "\n    ";
      for (int x = 0; x < width; x++) { cout << x%10;  } 
      cout << "\n";
	  
      cout << "----";
  	  for (int x = 0; x < width; x++) { cout << "-";  }
	  cout << "\n";

	  for (int y = 0; y < height; y++) {
	      cout << setw(2) <<  y%100 << ": ";
	      for (int x = 0; x < width; x++) {
              cout << board[y][x];
	      }
	      cout << "\n";
	  }
      
      cout << "----";
	  for (int x = 0; x < width; x++) { cout << "-";  }
	  cout << "\n";

	  cout << "    ";
	  for (int x = 0; x < width; x++) { cout << x/10;  } cout << "\n    ";
  	  for (int x = 0; x < width; x++) { cout << x/10;  }

	  cout << "\nFound " << countKnownMines() << " of " << totalMines << " mines\n";
  }

  void doReveal(int x, int y, int revealMines){
      for (int dy = -1; dy <=1 ; dy++) {
          for (int dx = -1; dx <=1 ; dx++) {
              int nx = x + dx;
              int ny = y + dy;
              if (in_bounds(nx,ny)) {
                  if (revealMines) {
                      //assert(board[ny][nx].square != UNKNOWN);
                      if (board[ny][nx].has_mine()){
                          board[ny][nx].mark_known_mine();
                      }
                  }
                  else  {
                      //assert(board[ny][nx].square != HAS_MINE);
                      if (board[ny][nx].is_unknown()) {
                          board[ny][nx].setCount(countMines(nx,ny));
                      }
                  }
              }
          }
      }
  }

  int maybeReveal(int x, int y) {
      int unknownSquares = 0;
      int knownMines = 0;
      for (int dy = -1; dy <=1 ; dy++) {
          for (int dx = -1; dx <=1 ; dx++) {
              int nx = x + dx;
              int ny = y + dy;
              if (in_bounds(nx,ny)) {
                  if (board[ny][nx].is_unknown()||
                      board[ny][nx].has_mine()) {
                      unknownSquares++;
                  }
                  else if(board[ny][nx].known_mine()) {
                      knownMines++;
                  }
              }
          }
    }
    assert(knownMines + unknownSquares >= board[y][x].count());
    assert(knownMines <= board[y][x].count());
    if (unknownSquares > 0) {
        int revealMines = (knownMines + unknownSquares == board[y][x].count());
        int allKnown = knownMines == board[y][x].count();
        if (revealMines || allKnown) {
            assert(!revealMines || !allKnown);
            doReveal(x,y, revealMines);
            return 1;
        }
    }
    return 0;
  }

  void determineKnownMines() {
      int foundMore = 0;
      for (int y = 0; y < height; y++) {
          for (int x = 0; x < width; x++) {
              if (board[y][x].count() >= 0) {
                  foundMore = maybeReveal(x,y) || foundMore;
              }
          }
      }
      if (foundMore) {
          determineKnownMines();
      }
  }

  void revealMines() {
      for (int y = 0; y < height; y++) {
          for (int x = 0; x < width; x++) {
              if (board[y][x].has_mine()) {
                  board[y][x].mark_known_mine();
              }
          }
      }
  }

  /**
   * returns true if the game is not over
   */
  bool playTurn() {
      cout << "Current board:\n";
      printBoard();
      cout << "x coordinate:\n";
      int x = readInteger();
      cout << "y coordinate:\n";
      int y = readInteger();

      if (!in_bounds(x,y)) {
          cout << "That is not a valid square, please try again\n";
      }
      else {
          ClickType result = board[y][x].click(countMines(x,y));
          determineKnownMines();
          if (result == CLICK_LOSE) {
              cout << "Oh no! That square had a mine. You lose!\n";
              revealMines();
              printBoard();
              return false;
          }
          else if (result == CLICK_KNOWN_MINE) {
              cout << "You already know there is a mine there!\n";
          }
          else if(checkWin()) {
              printBoard();
              cout << "You win!\n";
              return false;
          }
      }
      return true;
  }

  void playGame() {
      while (playTurn()) {
            ; // nothing to do, playTurn has all of the work, returns false when game over
      }
  }

};

bool checkPlayAgain() {
    string answer;
    cout << "Do you want to play again?\n";
    if (cin >> answer) {
        if (answer.length() > 0 && (answer[0] == 'Y' || answer[0] == 'y')) {
          return true;
        }
    }
    return false;
}

int main(int argc, char ** argv) {
    if (argc != 4) {
        cerr << "Usage: minesweeper width height numMines\n";
        return EXIT_FAILURE;
    }

    try {
        int width    = stoi(argv[1]);
        int height   = stoi(argv[2]);
        int numMines = stoi(argv[3]);
        if (width <= 0 || height <= 0 || numMines <= 0) {
            throw new std::invalid_argument("bad values");
        }

        do {
            Board b = Board(width, height, numMines);
            b.playGame();
        } while (checkPlayAgain());
    }
    catch(std::invalid_argument &a) {
       cerr << "Width, height, and numMines must all be positive ints\n";
       return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}

File: assignments/019-Minesweeper_py/minesweeper.py
# Implements the minesweeper game: https://en.wikipedia.org/wiki/Minesweeper_(video_game)
#
# As this assignments comes very early in your journey to learning Python,
# we have made some implementation decisions that are suboptimal as
# compared to using the entire language.

import random
import sys
from sys import argv

SQUARE_UNKNOWN = -1
SQUARE_HAS_MINE = -2
SQUARE_KNOWN_MINE = -3

CLICK_KNOWN_MINE = -2
CLICK_INVALID = -1
CLICK_CONTINUE = 0
CLICK_LOSE = 1


def board_width(board):
    return len(board[0])


def board_height(board):
    return len(board)


def add_random_mine(board):
    """ attempts to place a random mine onto the board """

    # maximum number of times to place a mine before given up
    max_tries = board_height(board) * board_width(board) * 10
    for i in range(max_tries):
        x = random.randint(0, board_width(board)-1)
        y = random.randint(0, board_height(board)-1)
        if board[y][x] == SQUARE_UNKNOWN:
            board[y][x] = SQUARE_HAS_MINE
            return


def make_board(width, height, num_mines):
    """
    Creates a minesweeper board using a list of lists.
    All entries are first created, setting each item to "SQUARE_UNKNOWN"
    Once that is done, "num_mines" are randomly placed on the board
    by calling add_random_mine repeatedly.
    """
    board = [[SQUARE_UNKNOWN for _ in range(width)] for _ in range(height)]
    
    for i in range(num_mines):
        add_random_mine(board)
    return board
    


def print_board(board):
    # print header
    print("   ", end="")
    for i in range(board_width(board)):
        print(i // 10, end="")
    print()
    print("   ", end="")
    for x in range(board_width(board)):
        print(x % 10, end="")
    print()
    print("-" * (board_width(board)+3))   # use string duplication

    for y in range(board_height(board)):
        print("{:2d}|".format(y), end="")
        for x in range(board_width(board)):
            if board[y][x] == SQUARE_UNKNOWN:
                print("?", end="")
            elif board[y][x] == SQUARE_HAS_MINE:
                print("?", end="")
            elif board[y][x] == SQUARE_KNOWN_MINE:
                print("*", end="")
            elif board[y][x] == 0:
                print(" ", end="")
            else:
                print(board[y][x], end="")
        print()


def has_mine(square):
    """ Returns true if a square has mine"""
    return square == SQUARE_HAS_MINE or square == SQUARE_KNOWN_MINE


def count_mines(board, x, y):
    """
    Counts the number of mines in the 8 squares around the (x,y) coordinate
    A mine is indicated by either SQUARE_HAS_MINE or SQUARE_KNOWN_MINE.
    We have provided a function that returns a Boolean value if a square has a
    mine.
    Make sure you do not go outside of the list bounds
    """
    num_mines = 0
    x_start, x_end, y_start, y_end = x, x, y, y
    x_start, x_end  = x - 1 if x > 0 else x_start ,   x + 1 if x < len(board[0]) - 1 else x_end
    y_start,  y_end = y - 1 if y > 0 else y_start,   y + 1 if y < len(board) - 1 else y_end
    
    for i in range(y_start, y_end + 1):
        for j in range(x_start, x_end + 1):
            if ( not(i == y and j == x) and has_mine(board[i][j])):
                num_mines += 1
    return num_mines

def click(board, x, y):
    if x < 0 or x > board_width(board) or y < 0 or y > board_height(board):
        return CLICK_INVALID
    elif board[y][x] == SQUARE_KNOWN_MINE:
        return CLICK_KNOWN_MINE
    elif board[y][x] == SQUARE_HAS_MINE:
        return CLICK_LOSE
    elif board[y][x] != SQUARE_UNKNOWN:
        return CLICK_CONTINUE
    else:
        board[y][x] = count_mines(board, x, y)
        return CLICK_CONTINUE


def check_win(board):
    """
    Game is won when no squares have the value SQUARE_UNKNOWN
    """
    for i in range(len(board)):
        for j in range(len(board[0])):
            if board[i][j] == SQUARE_UNKNOWN:
                return False
    return True

def reveal_mines(board):
    """
    Change SQUARE_HAS_MINE to SQUARE_KNOWN_MINE for all places on the board
    """
    for i in range(len(board)):
        for j in range(len(board[0])):
            if board[i][j] == SQUARE_HAS_MINE:
                board[i][j] = SQUARE_KNOWN_MINE
    return board

def read_int(prompt):
    while True:
        result = input(prompt)
        try:
            result = int(result)
            return result
        except Exception:
            print("Invalid entry")


def do_reveal(board, x, y):
    for nx in range(x - 1, x + 2):
        if nx < 0 or nx >= board_width(board):
            continue
        for ny in range(y - 1, y + 2):
            if ny < 0 or ny >= board_height(board):
                continue

            if board[ny][nx] == SQUARE_HAS_MINE:
                board[ny][nx] = SQUARE_KNOWN_MINE
            else:
                if (board[ny][nx] == SQUARE_UNKNOWN):
                    board[ny][nx] = count_mines(board, nx, ny)


def maybe_reveal(board, x, y):
    unknown_squares = 0
    known_mines = 0

    for nx in range(x - 1, x + 2):
        if nx < 0 or nx >= board_width(board):
            continue
        for ny in range(y - 1, y + 2):
            if ny < 0 or ny >= board_height(board):
                continue
            if board[ny][nx] == SQUARE_UNKNOWN or board[ny][nx] == SQUARE_HAS_MINE:
                unknown_squares += 1
            elif board[ny][nx] == SQUARE_KNOWN_MINE:
                known_mines += 1

    if unknown_squares > 0 and board[y][x] >= 0:
        reveal_mines = (known_mines + unknown_squares == board[y][x])
        all_known = known_mines == board[y][x]

        if reveal_mines or all_known:
            do_reveal(board, x, y)
            return True

    return False


def determine_known_mines(board):
    found_more = False
    for y in range(board_height(board)):
        for x in range(board_width(board)):
            if board[y][x] >= 0:
                found_more = maybe_reveal(board, x, y) or found_more
    if found_more:
        determine_known_mines(board)


def play_turn(board):
    """ plays a turn.  Returns True if the game is over, otherwise False"""
    print("Current Board:")
    print_board(board)

    x = read_int("x coordinate:")
    y = read_int("y coordinate:")

    result = click(board, x, y)
    determine_known_mines(board)

    if result == CLICK_LOSE:
        print("Oh no! That square had a mine. You lose!")
        reveal_mines(board)
        print_board(board)
        return True
    elif result == CLICK_INVALID:
        print("That is not a valid square, please try again")
        return False
    elif result == CLICK_KNOWN_MINE:
        print("You already know there is a mine there!\n")
        return False
    elif check_win(board):
        print_board(board)
        print("You win!")
        return True
    else:
        return False


# __name__ == "__main__" and argv explained in the "modules" notebook
if __name__ == "__main__":
    random.seed(42)    # seed the rng (set the beginning point) for validation]
    if len(argv) != 4:
        print("Usage: python3 minesweeper.py width height num_mines")
        sys.exit(-1)

    width = int(argv[1])
    height = int(argv[2])
    num_mines = int(argv[3])

    if (width <= 0 or height <= 0 or num_mines <= 0):
        print("Width, height, and num_mines must all be positive ints")
        sys.exit(-1)

    board = make_board(width, height, num_mines)
    game_over = False
    while not game_over:
        game_over = play_turn(board)
